tensordict/prototype/tensorclass.py

```diff
@@ -0,0 +1,357 @@
+                self.tensordict = _tensordict
+            else:
+                # get the default values
+                for key in datacls.__dataclass_fields__:
+                    default = datacls.__dataclass_fields__[key].default
+                    default_factory = datacls.__dataclass_fields__[key].default_factory
+                    if not isinstance(default_factory, dataclasses._MISSING_TYPE):
+                        default = default_factory
+                    if default is not None:
+                        kwargs.setdefault(key, default)
+
+                new_args = [None for _ in args]
+                new_kwargs = {
+                    key: None if key != "batch_size" else value
+                    for key, value in kwargs.items()
+                }
+                datacls.__init__(self, *new_args, **new_kwargs)
+
+                attributes = [key for key in self.__dict__ if key != "batch_size"]
+
+                for attr in attributes:
+                    if attr in dir(TensorDict):
+                        raise Exception(
+                            f"Attribute name {attr} can't be used for TensorDictClass"
+                        )
+
+                def _get_value(value):
+                    if isinstance(
+                        value, _accepted_classes + (dataclasses._MISSING_TYPE,)
+                    ):
+                        return value
+                    elif type(value) in CLASSES_DICT.values():
+                        return value.tensordict
+                    else:
+                        raise ValueError(f"{type(value)} is not an accepted class")
+
+                self.tensordict = TensorDict(
+                    {
+                        key: _get_value(value)
+                        for key, value in kwargs.items()
+                        if key not in ("batch_size",)
+                    },
+                    batch_size=kwargs["batch_size"],
+                )
+
+        @staticmethod
+        def _build_from_tensordict(tensordict):
+            return _TensorDictClass(_tensordict=tensordict)
+
+        @classmethod
+        def __torch_function__(
+            cls,
+            func: Callable,
+            types,
+            args: Tuple = (),
+            kwargs: Optional[dict] = None,
+        ) -> Callable:
+            if kwargs is None:
+                kwargs = {}
+            if func not in TD_HANDLED_FUNCTIONS or not all(
+                issubclass(t, (Tensor, cls)) for t in types
+            ):
+                return NotImplemented
+            return TD_HANDLED_FUNCTIONS[func](*args, **kwargs)
+
+        def __getattribute__(self, item):
+            def check_td_out_type(field_def):
+                if PY37:
+                    field_def = str(field_def)
+                if isinstance(field_def, str):
+                    if re.search(r"Optional\[.*\]", field_def):
+                        args = [re.search(r"Optional\[(.*?)\]", field_def).group(1)]
```

Comment:
These two searches are basically the same, might be better to do

```python
match = re.search(r"Optional\[(.*?)\]", field_def)
if match is not None:
    args = [match.group(1)]
```

We could also precompile these patterns at global scope say to save a little bit of work each time __getattribute__ is called

```python
# outside __getattribute__
OPTIONAL_PATTERN = re.compile(r"Optional\[(.*?)\]")

match = OPTIONAL_PATTERN.search(field_def)
```

Reply:
Want to take care of this @roccajoseph ?

Reply:
Yes, I'm on it.